src/groups/group.jl

```diff
@@ -0,0 +1,326 @@
+    return inverse_translate!(G, z, x, y, LeftAction())
+end
+
+"""
+    AdditionOperation <: AbstractGroupOperation
+
+Group operation that consists of simple addition.
+"""
+struct AdditionOperation <: AbstractGroupOperation end
+
++(e::Identity{G}) where {G<:AbstractGroupManifold{AdditionOperation}} = e
++(::Identity{G}, x) where {G<:AbstractGroupManifold{AdditionOperation}} = x
++(x, ::Identity{G}) where {G<:AbstractGroupManifold{AdditionOperation}} = x
++(e::E, ::E) where {G<:AbstractGroupManifold{AdditionOperation},E<:Identity{G}} = e
+
+-(e::Identity{G}) where {G<:AbstractGroupManifold{AdditionOperation}} = e
+
+function group_id!(::AbstractGroupManifold{AdditionOperation}, y, x)
+    copyto!(y, zero(x))
+    return y
+end
+
+group_id(::AbstractGroupManifold{AdditionOperation}, x) = zero(x)
+
+function inv!(::AbstractGroupManifold{AdditionOperation}, y, x)
+    copyto!(y, -x)
+    return y
+end
+
+inv(::AbstractGroupManifold{AdditionOperation}, x) = -x
+
+compose_left(::AbstractGroupManifold{AdditionOperation}, x, y) = x + y
+function compose_left!(::AbstractGroupManifold{AdditionOperation}, z, x, y)
+    z .= x .+ y
+    return z
+end
+
+
+"""
+    MultiplicationOperation <: AbstractGroupOperation
+
+Group operation that consists of multiplication.
+"""
+struct MultiplicationOperation <: AbstractGroupOperation end
+
+*(e::Identity{G}) where {G<:AbstractGroupManifold{MultiplicationOperation}} = e
+*(::Identity{G}, x) where {G<:AbstractGroupManifold{MultiplicationOperation}} = x
+*(x, ::Identity{G}) where {G<:AbstractGroupManifold{MultiplicationOperation}} = x
+*(e::E, ::E) where {G<:AbstractGroupManifold{MultiplicationOperation},E<:Identity{G}} = e
+
+/(x, ::Identity{G}) where {G<:AbstractGroupManifold{MultiplicationOperation}} = x
+/(::Identity{G}, x) where {G<:AbstractGroupManifold{MultiplicationOperation}} = inv(x)
+/(e::E, ::E) where {G<:AbstractGroupManifold{MultiplicationOperation},E<:Identity{G}} = e
+
+\(x, ::Identity{G}) where {G<:AbstractGroupManifold{MultiplicationOperation}} = inv(x)
+\(::Identity{G}, x) where {G<:AbstractGroupManifold{MultiplicationOperation}} = x
+\(e::E, ::E) where {G<:AbstractGroupManifold{MultiplicationOperation},E<:Identity{G}} = e
+
+# this is different from inv(G, e::Identity{G})
+inv(e::Identity{G}) where {G<:AbstractGroupManifold{MultiplicationOperation}} = e
+
+function group_id!(::AbstractGroupManifold{MultiplicationOperation}, y, x)
+    copyto!(y, one(x))
+    return y
+end
+
+group_id(::AbstractGroupManifold{MultiplicationOperation}, x) = one(x)
+
+function inv!(::AbstractGroupManifold{MultiplicationOperation}, y, x)
+    copyto!(y, inv(x))
+    return y
+end
+
+inv(::AbstractGroupManifold{MultiplicationOperation}, x) = inv(x)
+
+compose_left(::AbstractGroupManifold{MultiplicationOperation}, x, y) = x * y
+function compose_left!(::AbstractGroupManifold{MultiplicationOperation}, z, x, y)
+    #TODO: z might alias with x or y, we might be able to optimize it if it doesn't.
+    copyto!(z, x*y)
```

Comment:
Nice! I like how already this optimizes. e.g. if `x` is identity` and `y=z`, then this is a no op.

Reply:
Yes, sometimes it optimises but this should be non-allocating when `z` doesn't alias with `x` and `y`. I'll see if using `Base.dataids` would make sense.